test/gtest/ucp/test_ucp_request.cc

```diff
@@ -644,87 +638,101 @@ class test_proto_reset_atp : public test_proto_reset {
-            ucp_ep_get_lane(ep, lane)->iface->ops = m_ops[lane];
+            uct_ep_h uct_ep    = ucp_ep_get_lane(ep, lane);
+            uct_ep->iface->ops = m_ops[uct_ep];
         }
     }
 
+    static bool is_second_atp(uint8_t id)
+    {
+        static bool second = false;
+        bool res           = second && (id == UCP_AM_ID_RNDV_ATP);
+        second             = true;
+        return res;
+    }
+
+    static ucs_status_t uct_am_short(uct_ep_h ep, uint8_t id, uint64_t header,
+                                     const void *payload, unsigned length)
+    {
+        if (is_second_atp(id)) {
+            return UCS_ERR_NO_RESOURCE;
+        }
+
+        return m_ops[ep].ep_am_short(ep, id, header, payload, length);
+    }
+
     static ucs_status_t
     add_pending(uct_ep_h tl_ep, uct_pending_req_t *n, unsigned flag)
     {
+        m_reqs.push_back(n);
         return UCS_OK;
     }
 
     static void
     purge_pending(uct_ep_h ep, uct_pending_purge_callback_t cb, void *arg)
     {
-        if (m_req != NULL) {
-            cb(&m_req->send.uct, arg);
-            m_req = NULL;
+        for (auto &req : m_reqs) {
+            cb(req, arg);
         }
+
+        m_reqs.clear();
     }
 
 protected:
-    void wait_and_restart(const std::vector<request_pair_t> &pairs) override
+    void wait_and_restart(const std::vector<void*> &reqs) override
     {
-        ucp_ep_h ep                = sender().ep();
-        ucp_lane_index_t num_lanes = ucp_ep_config(ep)->key.num_lanes;
-
-        /* Backup uct ops for all lanes */
-        for (ucp_lane_index_t lane = 0; lane < num_lanes; ++lane) {
-            m_ops.push_back(ucp_ep_get_lane(ep, lane)->iface->ops);
-        }
-
         hook_uct_cbs();
 
-        /* Wait for rndv_put initialization */
-        auto &pair = wait_for_condition(pairs, [](const request_pair_t &pair) {
-            std::string rndv_put_zcopy_name("rndv/put/zcopy");
-            return rndv_put_zcopy_name ==
-                   pair.sreq->send.proto_config->proto->name;
-        });
-
-        restore_uct_cbs();
+        /* Wait until first ATP was sent and received by remote side */
+        static const unsigned send_stage = 3;
+        wait_any(reqs, [reqs](const void *ureq) {
+            if (ureq == NULL) {
+                return false;
+            }
 
-        /* Wait until ATP stage starts */
-        static const unsigned send_stage = 0;
-        const double timeout             = 10;
-        ucs_time_t deadline              = ucs::get_deadline(timeout);
+            ucp_request_t *req = (ucp_request_t*)ureq - 1;
+            if (!(req->flags & UCP_REQUEST_FLAG_PROTO_SEND)) {
+                return false;
+            }
 
-        while ((pair.sreq->send.proto_stage == send_stage) &&
-               (ucs_get_time() < deadline)) {
-            pair.sreq->send.uct.func(&pair.sreq->send.uct);
-        }
+            if (req->send.proto_stage != send_stage) {
```

Comment:
unite with line 694 using ||

Reply:
it's unsafe to unite them because if the request is recv request then `req->send.proto_stage` may conain garbage

Reply:
C defines a short circuit evaluation